src/antsibull/announcements.py

```diff
@@ -0,0 +1,151 @@
+# Copyright (C) 2024 Maxwell G <maxwell@gtmx.me>
+# SPDX-License-Identifier: GPL-3.0-or-later
+# GNU General Public License v3.0+
+# (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+"""
+Commands for generating release announcements
+"""
+
+from __future__ import annotations
+
+import asyncio
+import sys
+from collections.abc import Iterator
+from functools import partial
+from pathlib import Path
+from typing import TypedDict
+
+from aiohttp import ClientResponseError, ClientSession
+from antsibull_core import app_context
+from antsibull_core.dependency_files import DependencyFileData, DepsFile
+from jinja2 import Environment, PackageLoader, StrictUndefined, select_autoescape
+from packaging.version import Version as PypiVer
+
+from antsibull.constants import BUILD_DATA_URL
+from antsibull.pypi import PyPIClient, SdistAndWheel, UrlInfo
+
+ANNOUNCEMENTS = {
+    "ansible-email-announcement.txt": "ansible-email-announcement.j2",
+    "ansible-matrix-announcement.md": "ansible-matrix-announcement.j2",
+}
+
+eprint = partial(print, file=sys.stderr)
+
+jinja_env = Environment(
+    loader=PackageLoader(__package__, "data"),
+    autoescape=select_autoescape(),
+    trim_blocks=True,
+    undefined=StrictUndefined,
+)
+
+
+def email_heading(content):
+    """
+    Given a string, convert it into an email heading
+    """
+    return content + "\n" + "-" * len(content)
+
+
+jinja_env.filters["email_heading"] = email_heading
+
+
+class TemplateVars(TypedDict):
+    """
+    Variables used in the Jinja2 template
+    """
+
+    version: str
+    major_version: int
+    core_version: str
+    core_major_version: str
+    build_data_path: str
+    release_tarball: UrlInfo
+    release_wheel: UrlInfo
+
+
+def announcements_command() -> int:
+    """
+    Generate release announcements
+    """
+    app_ctx = app_context.app_ctx.get()
+    ansible_version: str = app_ctx.extra["ansible_version"]
+    output_dir: Path = app_ctx.extra["output_dir"]
+    dist_dir: Path | None = app_ctx.extra["dist_dir"]
+
+    deps_filename = Path(app_ctx.extra["data_dir"], app_ctx.extra["deps_file"])
+    deps_file = DepsFile(deps_filename)
+    dependency_data = deps_file.parse()
+    return asyncio.run(
+        _announcements_command(ansible_version, output_dir, dist_dir, dependency_data)
+    )
+
+
+async def verify_dists(dists: SdistAndWheel, dist_dir: Path) -> str | None:
+    for dist in dists:
+        file = dist_dir / dist.filename
+        if not await asyncio.to_thread(file.is_file):
```

Comment:
Wow, I didn't realize you don't need parens with `await` in such cases..

Anyway, I think this is hard to parse. Would you consider labeling this check with a human-readable var?

```suggestion

        dist_file_exists = await asyncio.to_thread(file.is_file)
        if not dist_file_exists:
```

I think, it'd be good for readability.

Reply:
I'm not a big fan of assigning an extra variable here. The reason I added the `asyncio.to_thread` thing is that `file.is_file` is a sync function that calls `stat()` which is technically blocking IO. I guess it doesn't really matter in this code, so I could just call the sync method directly.

Reply:
Async could persist, it's just that a named variable communicates the high-level intent of the check, which refers to the low-level details otherwise.

There's a difference between "an arbitrary file exists" and "a distribution is available". OTOH, I suppose this could be improved by renaming the `file` variable (which used to be a builtin, alias of `open()`, by the way).